File: Components/ASP.NET/CS/ASPxTreeListDemos/App_Code/NewsGroupDataProvider.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Web;
using System.Web.SessionState;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Linq;


public static class NewsGroupDataProvider  {
    static readonly string Key = "NewsGroupDataProvider";

	static HttpSessionState Session { get { return HttpContext.Current.Session; } }
    static List<Thread> Data {
		get {			
			if(Session[Key] == null)
				Restore();
            return (List<Thread>)Session[Key];
		}
	}	

	public static IEnumerable Select() {
		return Data;
	}

	public static void Insert(Thread item) {		
		Data.Add(item);
		item.ID = Data.Count;
	}

	public static void Update(Thread item) {
		Thread storedItem = FindItem(item.ID);
		Assign(item, storedItem);
	}

    private static void Assign(Thread source, Thread target) {
		target.ParentID = source.ParentID;
		target.From = source.From;
		target.Subject = source.Subject;
		target.Text = source.Text;
		target.DateCreated = source.DateCreated;
		target.HasAttachment = source.HasAttachment;
	}

	public static void Delete(Thread item) {
		Thread storedItem = FindItem(item.ID);
		Data.Remove(storedItem);
	}

	public static void Corrupt() {
		int count = 0;
		foreach(Thread item in Data) {
			int mod = count % 5;
			if(mod == 0)
				item.Subject = "";
			if(mod == 2)
				item.From = "";
			if(mod == 4)
				item.DateCreated = DateTime.Now.AddYears(-30 - count);
			count++;
		}
	}

	public static void Restore() {
		Session[Key] = CreateData();
	}

	static Thread FindItem(int id) {
        return Data.SingleOrDefault(t => t.ID == id);
	}

    static List<Thread> CreateData() {
        using(var context = new NewsGroupsContext())
            return context.Threads.ToList();
	}
}

File: Components/ASP.NET/VB/ASPxTreeListDemos/App_Code/NewsGroupDataProvider.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Data
Imports System.Web
Imports System.Web.SessionState
Imports System.Web.UI
Imports System.Web.UI.WebControls
Imports System.Linq


Public NotInheritable Class NewsGroupDataProvider
	Private Shared ReadOnly Key As String = "NewsGroupDataProvider"

	Private Sub New()
	End Sub
	Private Shared ReadOnly Property Session() As HttpSessionState
		Get
			Return HttpContext.Current.Session
		End Get
	End Property
	Private Shared ReadOnly Property Data() As List(Of Thread)
		Get
			If Session(Key) Is Nothing Then
				Restore()
			End If
			Return CType(Session(Key), List(Of Thread))
		End Get
	End Property

	Public Shared Function [Select]() As IEnumerable
		Return Data
	End Function

	Public Shared Sub Insert(ByVal item As Thread)
		Data.Add(item)
		item.ID = Data.Count
	End Sub

	Public Shared Sub Update(ByVal item As Thread)
		Dim storedItem As Thread = FindItem(item.ID)
		Assign(item, storedItem)
	End Sub

	Private Shared Sub Assign(ByVal source As Thread, ByVal target As Thread)
		target.ParentID = source.ParentID
		target.From = source.From
		target.Subject = source.Subject
		target.Text = source.Text
		target.DateCreated = source.DateCreated
		target.HasAttachment = source.HasAttachment
	End Sub

	Public Shared Sub Delete(ByVal item As Thread)
		Dim storedItem As Thread = FindItem(item.ID)
		Data.Remove(storedItem)
	End Sub

	Public Shared Sub Corrupt()
		Dim count As Integer = 0
		For Each item As Thread In Data
			Dim [mod] As Integer = count Mod 5
			If [mod] = 0 Then
				item.Subject = ""
			End If
			If [mod] = 2 Then
				item.From = ""
			End If
			If [mod] = 4 Then
				item.DateCreated = DateTime.Now.AddYears(-30 - count)
			End If
			count += 1
		Next item
	End Sub

	Public Shared Sub Restore()
		Session(Key) = CreateData()
	End Sub

	Private Shared Function FindItem(ByVal id As Integer) As Thread
		Return Data.SingleOrDefault(Function(t) t.ID = id)
	End Function

	Private Shared Function CreateData() As List(Of Thread)
		Using context = New NewsGroupsContext()
			Return context.Threads.ToList()
		End Using
	End Function
End Class
